CMakeLists.txt

```diff
@@ -401,6 +401,9 @@ add_library(${PROJECT_NAME}
 set_target_properties(${PROJECT_NAME} PROPERTIES
     VERSION 5.0.0
     SOVERSION 5)
+if ("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
+    target_link_options(${PROJECT_NAME} PRIVATE LINKER:--default-symver)
```

Comment:
This is failing CI because the CMake used on Travis-CI is too old to understand `target_link_options()`. Sorry, I don't know how to backport this (like I said, not a CMake expert).

Reply:
@smcv  try this

set_target_properties(${PROJECT_NAME} PROPERTIES
    VERSION 5.0.0
    SOVERSION 5
if("${CMAKE_VERSION}" VERSION_LESS "3.13.0" AND "${CMAKE_SYSTEM_NAME}" STREQUAL "Linux") 
    LINK_FLAGS -Wl,--default-symver )
elseif ("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
    )
	target_link_options(${PROJECT_NAME} PRIVATE LINKER:--default-symver)
else()
	)
endif()

Reply:
That looks like it'll just mis-compile the library with old cmake versions, please don't do that.

Also, this should not be hard coded to check for Linux, instead actually verify that the link flag is supported, perhaps like the check for `-Bsymbolic-functions` is handled.  That handling also is an example of how to include a link flag w/o the target_link_options (or add_link_options) support in newer cmake version.

Finally, rather than using `--default-symver`, as I mentioned on #621 I'd rather have a explicit version script.

Reply:
> instead actually verify that the link flag is supported

The reason I didn't take that approach is that if someone switches from a toolchain that does support it to a toolchain that doesn't, with a capability-based check they'll silently get a json-c whose ABI is incompatible with their previous build (going from versioned symbols to no versioned symbols is an incompatible change).

This is probably only a theoretical concern, because I tried a toolchain from 2012 and it already supported this; but if that downgrade does somehow happen, the question is whether you want a build failure that makes it obvious that something is wrong (as implemented here), or a successful build that silently drops the versioned symbols (which is what you'd get with a capability check).

If you're OK with that failure mode being possible, then yes I can take the same approach as for `-Bsymbolic-functions`.

Reply:
I think the opportunity to automatically include versioned symbols when the toolchain supports it, without needlessly limiting it to just Linux OSes, outweighs the chance of someone switching toolchains.

Reply:
OK, up to you. That approach is implemented in the latest version of this PR.